lib/datadog/appsec/contrib/rack/gateway/watcher.rb

```diff
@@ -56,7 +58,13 @@ def watch_request(gateway = Instrumentation.gateway)
 
                   next [nil, [[:block, event]]] if block
 
-                  ret, res = stack.call(request)
+                  ret, res = nil
+
+                  _, catched_request = catch(Datadog::AppSec::Ext::REQUEST_INTERRUPT) do
+                    ret, res = stack.call(request)
+                  end
+
+                  next [nil, catched_request] if catched_request
```

Comment:
Because since the event happen with the request flow, when we `throw` if we call `catch` in the `RequestMiddleware ` the code from the `watch_request` after calling `stack.call(request)` would continue and the result would be extra wrap the result in an extra array.

We can probably look for other places to have the `catch` located

Reply:
Thinking about it more, there are a bunch of things:

- `throw Datadog::AppSec::Ext::REQUEST_INTERRUPT`: `set_user` is theoretically and conceptually supposed to be called anytime and possibly not in a HTTP request. This is an argument to not have a name such as `REQUEST_INTERRUPT` in `set_user` (where I wanted it to be initially)
- the above also applies to the fact that the gateway controls the instrumentation and without appsec then there will be no throw by construction since there will be no watcher, just like that, and set_user is completely appsec-free
- yet I am wondering about:

  > I wanted this code to not be executed
  
  Is that what we want? kind of breaks the isolation of these watcher blocks
  
In any case I suggested some changes but now I feel this is actually going in the right direction, even though the latter point is kind of left up in the air for when we have more than one watch block.